File: ruby/lib/task_master/task.rb
module TaskMaster
  class Task
    attr_reader :id, :name, :assignee, :due_on, :status

    def initialize(id:, name:, assignee:, due_on:, status:)
      @id = id.to_i
      @name = name.to_s
      @assignee = assignee.to_s
      @due_on = Date.parse(due_on)
      @status = status.to_s
    end

    def incomplete?
      !complete?
    end

    def complete?
      status_of?("Completed")
    end

    def in_progress?
      status_of?("In Progress")
    end

    def not_started?
      status_of?("Not Started")
    end

    def status_of?(statuses)
      statuses.include?(status)
    end

    def assigned_to?(assignees)
      assignees.include?(assignee)
    end

    def upcoming?
      @due_on.between?(Date.today, next_week)
    end

    def overdue?
      @due_on < Date.today && incomplete?
    end

    def number_of_days_until_due
      (@due_on - Date.today).to_i
    end

    private

    def next_week
      Date.today + 7
    end
  end
end

File: javascript/lib/Task.js
class Task {
	constructor({id, name, assignee, dueOn, status}) {
		this.id = Number(id);
		this.name = String(name);
		this.assignee = String(assignee);
		this.dueOn = new Date(Date.parse(dueOn));
		this.status = String(status);
	}

	isIncomplete() {
		return !this.isComplete();
	}

	isComplete() {
		return this.hasStatusOf("Completed");
	}

	isInProgress() {
		return this.hasStatusOf("In Progress");
	}

	isNotStarted() {
		return this.hasStatusOf("Not Started");
	}

	hasStatusOf(statuses) {
		return [statuses].flat().includes(this.status);
	}

	isAssignedTo(assignees) {
		return [assignees].flat().includes(this.assignee);
	}

	isUpcoming() {
		const today = new Date()
		const nextWeek = new Date(today.getFullYear(), today.getMonth(), today.getDate() + 7)
		return this.#isDateBetween(this.dueOn, today, nextWeek)
	}

	isOverdue() {
		return this.dueOn < new Date() && this.isIncomplete();
	}

	numberOfDaysUntilDue() {
		return this.dueOn.getDate() - new Date().getDate()
	}

	#isDateBetween(targetDate, startDate, endDate) {
		return targetDate >= startDate && targetDate <= endDate;
	}
}

module.exports = Task
